cluster-api/apis/cluster/types.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package cluster contains types to represent Kubernetes cluster configuration.
+package cluster // import "k8s.io/kube-deploy/cluster-api/apis/cluster"
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// deepcopy-gen can be installed with:
+// go get k8s.io/gengo/examples/deepcopy-gen
+//go:generate deepcopy-gen -i . -O zz_generated.deepcopy
+
+// Cluster is an API object representing cluster-wide configuration
+// parameters and status.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Cluster struct {
+	metav1.ObjectMeta `json:"metadata"`
+	metav1.TypeMeta   `json:",inline"`
+
+	Spec   ClusterSpec   `json:"spec"`
+	Status ClusterStatus `json:"status,omitempty"`
+}
+
+type ClusterSpec struct {
+	// Cluster network configuration
+	ClusterNetwork ClusterNetworkingConfig `json:"clusterNetwork"`
+
+	// Provider-specific serialized configuration to use during
+	// cluster creation. It is recommended that providers maintain
+	// their own versioned API types that should be
+	// serialized/deserialized from this field.
+	ProviderConfig string `json:"providerConfig"`
+}
+
+// ClusterNetworkingConfig specifies the different networking
+// parameters for a cluster.
+type ClusterNetworkingConfig struct {
+	// The subnet CIDR from which service VIPs are allocated.
+	ServiceSubnet string `json:"serviceSubnet"`
+
+	// The subnet CIDR from which POD networks are allocated.
+	PodSubnet string `json:"podSubnet"`
+
+	// Domain name for services.
+	DNSDomain string `json:"dnsDomain"`
+}
+
+// ClusterStatus represents the current status of the cluster.
+type ClusterStatus struct {
+	// APIEndpoint represents the endpoint to communicate with the IP.
+	APIEndpoints []APIEndpoint `json:"apiEndpoints"`
+
+	// If set, indicates that there is a problem reconciling the
```

Comment:
Shouldn't we try to group this into something similar to NodeCondition? Right now its hard to tell for example when the error was thrown.

Reply:
@pipejakob I remember you got some feedback on the machines portion about not using conditions. Could you comment on that?

Reply:
See https://github.com/kubernetes/kubernetes/issues/7856#issuecomment-335687733